### DISCUSSION TITLE
Rendering issue when in release mode or debugger is not attached

### DISCUSSION THREAD

**Discussion by SeanFeldman** [gismofx/toast_ui.blazor_calendar#36] [Q&A] [answered]
This is likely something I'm missing but for the life of me can't figure out what is it.

When running with the debugger attached, everything is working great and renders fine. The moment running w/o debugger attached or in release mode, the rendering gets messed up. The data should contain a single schedule that starts and end on the same date.
<img src="https://user-images.githubusercontent.com/1309622/171549086-f2b5d864-db6b-4497-87b4-d59b605f9682.png" width="400" />

When looking at the popup, I'm noticing some weirdness in the data that otherwise is valid data when received from REST API.
<img width="240" alt="image" src="https://user-images.githubusercontent.com/1309622/171549480-8c656766-85bc-4995-b05f-6f0a0d49faa2.png">

Any ideas on what could cause it? Or suggestions on how to debug it?

PS: I'm using the control with Blazor WebAssembly

**Comment by gismofx** [comment, marked as answer]
My first thought is the mapping between the incoming data to a TUISchedule is not behaving. Using VS2022? Can you confirm your incoming data is correct? Could you share your project or make a simplified version that re-produces the issue?.. I'd be glad to take a look.

**Reply by SeanFeldman** [reply]
This is happening when deployed. That’s how I found out the difference. When tried running w/o debugger attached locally (VS2022), could see the same. The data is the same data that renders fine when debugger is attached. So I ruled that out.
I’ll try put a stand-alone repro to see if can get the same issue to repeat.

**Reply by SeanFeldman** [reply]
@gismofx I haven't had a chance to get a repro done but I did briefly look into what's returned from the server-side in release mode. It looks the same as in debug mode:

```json
[
  {
    "RequestId": "cc781243-7095-4d94-abb7-b0e3579d46b9",
    "RequestTypeName": "---redacted---",
    "AppointmentDateTime": "2022-05-26T08:00:00",
    "GmtOffset": -18000,
    "AppointmentTimeZone": "CDT",
    "InsuredAddress": "---redacted---",
    "InsuredCity": "---redacted---",
    "InsuredRegion": "---redacted---",
    "InsuredPostalCode": "---redacted---",
    "InsuredCountry": "---redacted---",
    "Name": "---redacted---",
    "ClientDescriptor": "---redacted---"
  }
]
```

I've also taken a peek at what's passed down to the JavaScript but not sure if it's valid or not:
<image src="https://user-images.githubusercontent.com/1309622/172083321-b383f4e1-924a-44ba-868c-fbcff54225ba.png" width=300/>

But the calendar is failing to render those `start` and `end` dates
![image](https://user-images.githubusercontent.com/1309622/172084182-ed22bffe-9e68-4bea-b88d-5c860c4cf392.png)

**Reply by gismofx** [reply]
Where is the duration coming from? How are you mapping/creating the TUISchedule which gets sent to the browser?

**Reply by SeanFeldman** [reply]
The start and end dates are coming from .NET code. Here it is:
```c#
async Task OnSelectedValuesChanged(IEnumerable<Tva.Shared.User> values)
    {
        updatingCalendar = true;

        var consultations = await RequestsConsultationService.GetRequestsWithScheduledAppointment(values.First().UserId);

        var schedules = consultations.Select(x => new TUISchedule
        {
            id = x.RequestId.ToString(),
            calendarId = "1",
            start = x.AppointmentDateTime,
            end = x.End,
            title = x.Title,
            body = x.Body,
            attendees = new[] { x.Name },
            location = x.Location,
            category = "time",
            isVisible = true,
            isAllDay = false,
            state = "busy",
            //bgColor = "#3a87ad",
            color = "#ffffff"
        });

        await ClearCalendar();

        await calendarRef.CalendarInterop.CreateSchedulesAsync(schedules);

        updatingCalendar = false;
    }

`AppointmentDateTime` and `End` are both `DateTimeOffset`
```

**Reply by gismofx** [reply]
I'm wondering if this is related to #35 with time zones in wasm. I think the main issue sense that there is an issue between debug vs deployed since they are compiled differently.

Do the dates show up correctly in the edit popup? Are they correctly placed on the calendar. If you change the time in the calendar, does it change the display in the popover?

Are you using Month View or week view, both?

Issue could be here and need to dig a little more(Line 60): https://github.com/gismofx/toast_ui.blazor_calendar/blob/6ff1d688b3c32b12633a07edf437f7edd76c7838/toast_ui.blazor_calendar/Services/JsonConverters/TZDateJsonConverter.cs#L56-L60

Here's the format from the TUI API. Maybe need to customize how the data is formatted when serialized and sent to JS. I can try to play with this
```js
calendar.createSchedules([
    {
        id: '1',
        calendarId: '1',
        title: 'my schedule',
        category: 'time',
        dueDateClass: '',
        start: '2018-01-18T22:30:00+09:00',
        end: '2018-01-19T02:30:00+09:00'
    },
    {
        id: '2',
        calendarId: '1',
        title: 'second schedule',
        category: 'time',
        dueDateClass: '',
        start: '2018-01-18T17:30:00+09:00',
        end: '2018-01-19T17:31:00+09:00'
    }
]);
```

**Reply by gismofx** [reply]
@SeanFeldman I pushed up one change to the Dev branch. Can you try and let me know if that fixes it for you?
https://github.com/gismofx/toast_ui.blazor_calendar/blob/cf3901f40556e1d9c54d1f6aa61f99d5fa5f1cae/toast_ui.blazor_calendar/Services/JsonConverters/TZDateJsonConverter.cs#L60

**Reply by SeanFeldman** [reply]
@gismofx you're not pushing the preview package anywhere, are you?
Would be nice to have automation in place to build the package from `master` as well as a manual kicked off the pipeline to build preview packages for occasions such as this.

**Reply by SeanFeldman** [reply]
The serialization change did fix the issue and now it renders in all 3 modes:

<img width="412" alt="image" src="https://user-images.githubusercontent.com/1309622/172985584-8e2cd890-3d59-4776-b679-4df5b2bd3efa.png">

<img width="215" alt="image" src="https://user-images.githubusercontent.com/1309622/172985774-b8e13542-b4f0-4163-98dd-d133e846a4d9.png">

<img width="316" alt="image" src="https://user-images.githubusercontent.com/1309622/172985933-d72be346-14aa-4665-9731-a256166581c3.png">

Hopefully, you can push out an updated version.

**Reply by gismofx** [reply]
Yay! Glad that's fixed. I should be able to push an updated nuget in the next day or so. 



> @gismofx you're not pushing the preview package anywhere, are you?
> Would be nice to have automation in place to build the package from `master` as well as a manual kicked off the pipeline to build preview packages for occasions such as this.

Yes, I absolutely agree, but I need some help with more of that build automation...or a push in the right direction. Have any resources/docs on how to set it up properly in this case?

**Reply by SeanFeldman** [reply]
I could give a hand if you'd like. A small token of appreciation for your work 🙂

**Reply by gismofx** [reply]
I would be happy to accept (and learn!) Thank you. 

I'll create a new issue.

**Reply by SeanFeldman** [reply]
Looking forward to the release. Thanks.

**Reply by gismofx** [reply]
@SeanFeldman the fix was merged into main and Nuget package updated! Let me know how it goes.

**Reply by SeanFeldman** [reply]
Working great. Thank you, @gismofx for such a quick update.

Next is #38

**Reply by gismofx** [reply]
Of course! Thanks for feedback and patience. Can you mark this discussion as answered?
